src/flow/src/expr/scalar.rs

```diff
@@ -428,6 +261,89 @@ impl ScalarExpr {
         }
     }
 
+    pub fn eval_batch(&self, batch: &Batch) -> Result<VectorRef, EvalError> {
+        match self {
+            ScalarExpr::Column(i) => Ok(batch.batch()[*i].clone()),
+            ScalarExpr::Literal(val, dt) => Ok(Helper::try_from_scalar_value(
+                val.try_to_scalar_value(dt).context(DataTypeSnafu {
+                    msg: "Failed to convert literal to scalar value",
+                })?,
+                batch.row_count(),
+            )
+            .context(DataTypeSnafu {
+                msg: "Failed to convert scalar value to vector ref when parsing literal",
+            })?),
+            ScalarExpr::CallUnmaterializable(_) => OptimizeSnafu {
+                reason: "Can't eval unmaterializable function",
+            }
+            .fail()?,
+            ScalarExpr::CallUnary { func, expr } => func.eval_batch(batch, expr),
+            ScalarExpr::CallBinary { func, expr1, expr2 } => func.eval_batch(batch, expr1, expr2),
+            ScalarExpr::CallVariadic { func, exprs } => func.eval_batch(batch, exprs),
+            ScalarExpr::CallDf {
+                df_scalar_fn,
+                exprs,
+            } => df_scalar_fn.eval_batch(batch, exprs),
+            ScalarExpr::If { cond, then, els } => {
+                let conds = cond.eval_batch(batch)?;
+                let bool_conds = conds.as_any().downcast_ref::<BooleanArray>().context({
+                    TypeMismatchSnafu {
+                        expected: ConcreteDataType::boolean_datatype(),
+                        actual: conds.data_type(),
+                    }
+                })?;
+
+                let mut ret_vec = None;
+
+                for (idx, cond) in bool_conds.iter().enumerate() {
+                    let input_vectors = batch
+                        .batch()
+                        .iter()
+                        .map(|v| v.clone().slice(idx, 1))
+                        .collect::<Vec<_>>();
+                    let input_batch = Batch::new(input_vectors, 1);
+                    let res: VectorRef = match cond {
+                        Some(true) => then.eval_batch(&input_batch)?,
+                        Some(false) => els.eval_batch(&input_batch)?,
+                        None => Arc::new(NullVector::new(1)),
+                    };
```

Comment:
Either call `eval` for one value directly, or separate the values ​​whose cond is true/false/none and call `eval_batch` and then assemble them back?